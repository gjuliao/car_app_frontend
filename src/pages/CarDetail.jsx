/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams, Link } from 'react-router-dom';
import { getCar } from '../redux/carDetail/carDetailSlice';
import DetailTable from '../components/DetailTable';
import styles from '../styles/CarDetail.module.css';

export default function Detail() {
  const dispatch = useDispatch();
  const user = useSelector((store) => store.session.data);
  const data = useSelector((store) => store.car.data);
  const { id } = useParams();

  useEffect(() => {
    dispatch(getCar(id));
  }, []);

  return (
    <>
      {data.errors === true ? (
        <div className={styles.errorContainer}>
          <p>{data.message || 'Error'}</p>
        </div>
      ) : (
        <section className={styles.container}>
          <div className={styles.photoArea}>
            <img className={styles.photo} src={data.payload?.image} alt="car" />
          </div>
          <div className={styles.infoArea}>
            <div className={styles.features}>
              <h1 className={styles.title}>{data.payload?.model}</h1>
              <p className={styles.subtitle}>
                {`- $${data.payload?.price * 0.1} deposit upon any ${data.payload?.model} purchase`}
              </p>
              <DetailTable data={data.payload} />
            </div>

            {Object.keys(user).length === 0 ? (
              <p>You need to be logged in to reserve a car</p>
            ) : (null)}

            {user.payload?.role === 'admin' ? (
              <p>Administrators can not reserve a car</p>
            ) : (null)}

            {Object.keys(user).length !== 0 && user.payload?.role !== 'admin' ? (
              <Link to={`/reservation/${id}`}>
                <button type="button" className={styles.reserveBtn}>Reserve</button>
              </Link>
            ) : (null)}

          </div>
        </section>
      )}
    </>
  );
}
