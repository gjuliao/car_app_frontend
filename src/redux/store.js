import { configureStore } from '@reduxjs/toolkit';
import reservationReducer from './reservation/reservationSlice';
import sessionReducer from './sessionSlice';
import carListReducer from './carlistSlice';
import carReducer from './carDetail/carDetailSlice';
import carItemsReducer, { fetchCars } from './car/carSlice';
import auth from './Auth/auth';

const store = configureStore({
  reducer: {
    auth,
    reservations: reservationReducer,
    session: sessionReducer,
    cars: carListReducer,
    car: carReducer,
    carlist: carItemsReducer,
  },
});
store.dispatch(fetchCars());
export default store;
