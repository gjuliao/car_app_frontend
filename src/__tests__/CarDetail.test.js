import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import store from '../redux/store';
import CarDetail from '../pages/CarDetail';

describe('CarDetail page', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <CarDetail />
        </BrowserRouter>
      </Provider>,
    );
  });

  test('Show a message when the user is not logged in', () => {
    expect(screen.getByText('You need to be logged in to reserve a car')).toBeDefined();
  });

  test('AddCar renders without crashing', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <CarDetail />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
