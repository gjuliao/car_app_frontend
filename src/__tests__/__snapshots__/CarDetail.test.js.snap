// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarDetail page AddCar renders without crashing 1`] = `
<section
  className="container"
>
  <div
    className="photoArea"
  >
    <img
      alt="car"
      className="photo"
    />
  </div>
  <div
    className="infoArea"
  >
    <div
      className="features"
    >
      <h1
        className="title"
      />
      <p
        className="subtitle"
      >
        - $NaN deposit upon any undefined purchase
      </p>
      <table
        className="detailTable"
      >
        <tbody>
          <tr>
            <td
              className="keyRow"
            >
              Price
            </td>
            <td
              className="valueRow"
            >
              $
              Loding...
            </td>
          </tr>
          <tr>
            <td
              className="keyRow"
            >
              Brand
            </td>
            <td
              className="valueRow"
            >
              Loding...
            </td>
          </tr>
          <tr>
            <td
              className="keyRow"
            >
              Year
            </td>
            <td
              className="valueRow"
            >
              Loding...
            </td>
          </tr>
          <tr>
            <td
              className="keyRow"
            >
              Color
            </td>
            <td
              className="valueRow"
            >
              Loding...
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <p>
      You need to be logged in to reserve a car
    </p>
  </div>
</section>
`;
